### PR TITLE
Documentation for Phinx via Code added

### PR DESCRIPTION
**Pull request by alexbowers**
Documentation Added for calling Phinx via code.

### CODE REVIEW

File: docs/code-usage.rst [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    the best way to manage this would be to ``require "vendor/autoload.php";`` the composer autoloader.
+
+Phinx will require instantiating, which can be achieved by the following.
+
+.. code-block:: php
+
+    $app = new \Phinx\Console\PhinxApplication($version);
+    $phinx = new \Phinx\Wrapper\TextWrapper($app);
+
+Where ``$version`` is the version number of Phinx (eg. '0.4.1').
```
Comment (shadowhand) [inline docs/code-usage.rst]: This works, but it might be better to use:

```php
$app = require 'vendor/robmorgan/phinx/app/phinx.php';
```

File: docs/code-usage.rst [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    If left blank, the value is assumed based on the Configuration File's file extension as per the list above.
+
+Migrating
+----------
+
+This command will run the same command as ``phinx migrate``.
+
+.. code-block:: php
+
+    $phinx->getMigrate($env, $target);
```
Comment (shadowhand) [inline docs/code-usage.rst]: I think this example should include the return value:

```php
$output = $phinx->getMigrate(...);
```

File: docs/code-usage.rst (hunk 1 above)
Comment (alexbowers) [inline docs/code-usage.rst]: I have to disagree.

Many people will already have composer's autoloading defined, and thus requiring phinx.php will autoload PhinxApplication twice.

Also, I think this:
    $app= new PhinxApplication();
is clearer than
    $app= require 'vendor/robmorgan/phinx/app/phinx.php';
because its immediately clear whats going on.

Comment (alexbowers) [inline docs/code-usage.rst]: I just saw your edit.

I agree that having an embedded one would be nice actually. Though, to me it would still be nicer if we had a Facade style class which was simply called Phinx which wrapped the TextWrapper and some of its methods

Having Phinx->migrate() is clearer than TextWrapper->getMigrate()

File: docs/code-usage.rst (hunk 2 above)
Comment (alexbowers) [inline docs/code-usage.rst]: Wouldn't the most common use case of Phinx via Code be without needing the return value for the migration. Since it is meaningless to the end-user (who is either running an install script, or something).

I can add it if you want, but the use case is more limited than without it.

Comment (shadowhand) [inline docs/code-usage.rst]: The above comment and this comment make sense... Some modifications could be made that would clean all this up:

1. `$phinx = new Phinx\Phinx()` to get a fluent object
2. `$phinx->migrate($target = null)` and `$phinx->rollback($target = null)` that return a  boolean (based on exit code)
3. updated documentation

In many ways, this points back #413 and suggests that Phinx should be internally refactored to exist as a library, which also provides a console application that **uses** the library, rather than having the opposite situation (as it is currently).

Comment (alexbowers) [inline docs/code-usage.rst]: To me, what makes sense is to have 1 and 2 as you specified. However, in a way that the message is also split down and segmented into an array of messages and timestamps etc, so that if the user wished to get the response, such as 'xxx migrated 0.2s' they can simply call a method like $phinx->getMessage(); and it will have the same message as $phinx->migrate() currently has, and with work done on the message too, we can provide it in many formats / split up.

I have to agree with the use of a library rather than a command.

Comment (alexbowers) [inline docs/code-usage.rst]: Shall we commit it as is for now, and have an RFC for refactoring into a library?

Comment (alexbowers) [inline docs/code-usage.rst]: @shadowhand Shall we open this up as a RFC?

Comment (shadowhand) [inline docs/code-usage.rst]: I'd like to see the code and the documentation together in a single PR.

Comment (alexbowers) [inline docs/code-usage.rst]: Hasn't the code already been merged in?

Comment (shadowhand) [inline docs/code-usage.rst]: Not the fluent object we're talking about.

Comment (alexbowers) [inline docs/code-usage.rst]: Ah sorry. I thought you were meant that was a future change. I'll start
work on that tonight

Comment (shadowhand) [inline docs/code-usage.rst]: Have you made any progress on the fluent class, or should I start working on it?

Comment (alexbowers) [inline docs/code-usage.rst]: I haven't had chance to start yet. If you can make it a branch I may have time to contribute a bit to it tonight, but i've started on a PoC for the company I work for.

**Comment by shadowhand** [comment]
@alexbowers check out https://github.com/robmorgan/phinx/tree/i413-fluent-phinx

**Comment by shadowhand** [comment]
That's only a basic starting point... I'd actually like to see a bigger refactor project that inverts Phinx more completely and makes the fluent interface the core part, and the console application a wrapper.

**Comment by alexbowers** [comment]
I don't seem to be able to add comments to those files.

But a few thoughts from me so far:

All numbers refer to line numbers:

64: I prefer to have paths which have ../ ../ in them to be `dirname(dirname(__DIR__)). That way, when debugging, we get the actual path, rather than a convoluted way to get to it. 

97, 110: Not a particarly nice set of parameters there, passing null through first. Maybe TextWrapper could be refactored to remove them since we can't pass through values to it anymore?

**Comment by alexbowers** [comment]
@shadowhand Is there anything else required for a merge?

Also, the documentation requires updating again. I'm a little busy at the minute, so if you wouldn't mind could you please update the docs with the new code examples and PR them into my branch?

If not, I'll probably get around to them tomorrow.

**Comment by alexbowers** [comment]
@shadowhand Documentation has been updated.

**Comment by alexbowers** [comment]
@shadowhand Any idea when this will be merged in along with Fluent.

I have an application which will require updating when this is merged; and having it done sooner rather than later would be nice.

**Comment by shadowhand** [comment]
@alexbowers honestly, creating the Fluent interface doesn't make a lot of sense without the refactor, which is a large task and cause the code to diverge significantly. There are already multiple ways to work around this without the fluent interface in your application, or you can import the existing class I created into your project.

tl;dr: there are no plans to add the fluent interface soon, handle it in your own project for now.

**Comment by alexbowers** [comment]
Okay. I was going to look into the refractor this evening. Is it not worth
me looking into it yet?

**Comment by shadowhand** [comment]
@alexbowers I'm not opposed to you trying, but it requires significant understand of the codebase and how the 3 parts (manager, env, migrations) all work together.

**Comment by alexbowers** [comment]
@shadowhand Any chance of getting this docs merged in too?

**Comment by shadowhand** [comment]
@alexbowers honestly I would prefer not to document/encourage usage of the current interfaces until the underlying architecture is improved. Even the current names are pretty bad.

**Comment by alexbowers** [comment]
Okay. can we draft up a RFC at some point if we're heading towards backwards incompatibility?

That will also allow some of us to fork off and create a new architecture

**Comment by robmorgan** [comment]
I'm closing this. The current interfaces should not be documented as is.
